WIP: Ad-hoc observables for grains (#940)

ReubenBond opened the pull request:
Work-in-progress of ad-hoc observables for grains, as described in #940.

## Edit:
The API is now aligned with existing Virtual Streams, so grains can directly return `IAsyncObservable<T>` instead of the custom `IGrainObserver<T>` type I originally had here. This means that `ResumeAsync` works. Discussed more in [this comment](https://github.com/dotnet/orleans/pull/1916#issuecomment-232331824) below.

## Original:
This allows users to define grain interfaces like so:
```C#
public interface IReactiveGrain<T> : IGrainWithGuidKey
{
    IGrainObservable<T> GetStream(T[] values);
}
```
```C#
public class ReactiveGrain<T> : Grain, IReactiveGrain<T>
{
    public IGrainObservable<T> GetStream(T[] values)
    {
        return Observable.Create<T>(async observer =>
        {
            if (values == null) throw new ArgumentNullException(nameof(values));
            try
            {
                foreach (var value in values) await observer.OnNext(value);
                await observer.OnCompleted();
            }
            catch (Exception exception)
            {
                await observer.OnError(exception);
            }

            return new AsyncDisposable();
        });
    }
}
```

More advanced usage could use filter values or expression trees as parameters.

There are obvious deficiencies, some of which are called out in the code. For example, observables should error or be disposed when the remote side dies (depending on which side dies). ~~Additionally, this only works for `int` at the moment. It's an easy fix with a bit of IL.~~ Please raise any concerns you have.

This is definitely not a replacement for Virtual Streams, this is a complimentary feature.

The remote grain is only called when the returned observable is subscribed to, it is called every time `Subscribe` is called. The observable captures the arguments at the time of creation so that it can subscribe and resubscribe as needed.

Let me know what you think - all criticism is welcome.

gabikliot commented:
I am probably missing something, but two questions: why do u you need the new interface I grain observarable? We already have I async observarable.
Also, I don't understand the sample code. What is the array argument? How does the full code usage example works? I presume there are answers to both questions in the pr, but I didnt look in it yet. Only basing my questions on the issue comment.

ReubenBond commented:
> why do u you need the new interface I grain observarable? We already have I async observarable.

These are placeholder interfaces, but I avoided using the in-built `IAsyncObservable` because `Subscribe` takes a bunch of unneccessary parameters and returns a `StreamSubscriptionHandle<T>`. I can use those anyway, but I'm not sure what to fill those values with. Similarly for `IAsyncObserver<T>`.

> Also, I don't understand the sample code. What is the array argument? How does the full code usage example works?

The example code just returns the values you pass it, one-by-one. It's completely contrived & was written for testing purposes. The grain could instead hold on to the observer and pass values into it later on.

ReubenBond commented:
> we probably will not want to have a separate set of interfaces that logically express the same, right?

Those interfaces have parameters which are for a more sophisticated system. For example, in addition to `UnsubscribeAsync`, `StreamSubscriptionHandle<T>` has `ResumeAsync`, `Equals`, `HandleId`, and `StreamIdentity`, which may make sense for Virtual Streams, which are often durable, one-to-many or many-to-many, and generally much more advanced, but they don't make so much sense for ad-hoc streams.
I'm playing around with it some more anyhow.

> For prototyping what you put on is fine, but long term programming model, I am afraid the programming model police (@sergeybykov) would not approve it. ;-)

That's why I'm opening the issue now: so discussion can shape the PR as it evolves :)

ReubenBond commented:
Pushed an update which adds support for observables of types other than `int`. It uses IL to optimize calls to the subscribe method (prematurely, I'm sure)

AlgorithmsAreCool commented:
Take a look at the experimental interfaces defined in corefxlab for a potential design 

https://github.com/dotnet/corefxlab/tree/master/src/System.Threading.Tasks.Channels

jdom commented:
Hehe, cool that we are exploring this. I personally would prefer that if we go with IObservable, we use the framework APIs, meaning that the producer will just send the data (and not await any task, as getting feedback is not part of IObservable in the first place). This shouldn't be completely fire and forget, the consumer should either receive ALL of the messages in order or get OnError and the observer subscription terminated.
The consumer is a tougher thing to solve (but doable), because with IObserver data should be processed either with a synchronous function or you can explicitly have helpers that allow you to serialize asynchronous processing. We could have such a helper that honors the grain reentrancy settings, and treat each OnNext as a new grain request (but this would be of course opt in, the grain can decide to process synchronously).
Didn't dig too much, and I'm no expert in Rx, but something similar to [this bridge](http://stackoverflow.com/a/35190003) is the kind of thing we can implement to bridge sync methods to grains' threading model. Maybe there is something already built into the new version of Rx.
The key thing is that once we talk the Rx language explicitly (not our own API which kind of resembles it), we can get the benefit of all the APIs already built around it, and the familiarity and understanding with a well documented library with a big community (such as there are subtle semantics for how observers should work, that if we implement our own version, then we might start breaking those semantics because we have nothing to conform to).

BTW, the Channels stuff that @AlgorithmsAreCool mentioned is also cool for stuff that is inherently async, as a simpler API than DataFlow. DataFlow was the other alternative I was going to propose if we decide to go fully async (I already successfully used DataFlow in Halo 5 to turn our use of IGrainObserver into something that can more easily be used in a connected client scenario using AMQP over websockets, but still having proper Orleans support for it can improve it in the sense that ordering can be enforced, otherwise complete the block with failure, and also avoid/optimise our constant keep alive pings to know when either the producer or consumer vanished due to failure).

My 2 cents, don't want to completely derail this PR with discussion, as this is an idea I had but didn't really try to implement it to find roadblocks that we are probably likely to encounter. We can continue talking over Gitter.

ReubenBond commented:
Thank you for the input @AlgorithmsAreCool & @jdom.
So our options are:
* `IReadableChannel<T>` / `IWriteableChannel<T>` from the .NET CoreFx' incubation repo
* `IObservable<T>` / `IObserver<T>` from System.Reactive / Rx
* `ISourceBlock<T>` / `ITargetBlock<T>` from TPL Dataflow
* `IAsyncObservable<T>` / `IAsyncObserver<T>` from Orleans Virtual Streams
* `IGrainObservable<T>` / `IGrainObserver<T>` from this WIP PR

We can choose more than one option, but each brings with it implementation cost and the cost of confusing end users.

In the latest version of the PR I've added support for grains subscribing to other grains (which made use of a little dependency injection). I also added a silly `ChatRoomGrain` & `ChatUserGrain` + tests to demonstrate how this can be used.
Here's `ChatRoomGrain`:
```c#
public class ChatRoomGrain : Grain, IChatRoomGrain
{
    private readonly MulticastObservable<string> room = new MulticastObservable<string>();
    public IGrainObservable<string> JoinRoom() => room;
    public Task SendMessage(string message) => room.OnNext(message);
    public Task<int> GetCurrentUserCount() => Task.FromResult(room.Count);
}
```

AlgorithmsAreCool commented:
For convenience, here are the API surfaces that @ReubenBond just mentioned.

I have subjectively ordered them by how appropriate i think they are.

### 1. Current Design

```C#
public interface IGrainObserver<in T>
{
	Task OnNext(T value);
	Task OnError(Exception exception);
	Task OnCompleted();
}

public interface IGrainObservable<out T>
{
	Task<IAsyncDisposable> Subscribe(IGrainObserver<T> observer);
}
```

### 2. Orleans Virtual Streams

```C#
public interface IAsyncObservable<T>
{
	Task<StreamSubscriptionHandle<T>> SubscribeAsync(IAsyncObserver<T> observer);

	Task<StreamSubscriptionHandle<T>> SubscribeAsync(IAsyncObserver<T> observer, StreamSequenceToken token, StreamFilterPredicate filterFunc = null, object filterData = null);
}

public interface IAsyncObserver<in T>
{
	Task OnNextAsync(T item, StreamSequenceToken token = null);

	Task OnCompletedAsync();

	Task OnErrorAsync(Exception ex);
}
```


### 3. CoreFxLab Channels (Experimental)
[System.Threading.Tasks.Channels](https://github.com/dotnet/corefxlab/tree/master/src/System.Threading.Tasks.Channels)

```C#
public interface IReadableChannel<T>
{
    bool TryRead(out T item);
    ValueTask<T> ReadAsync(CancellationToken cancellationToken = default(CancellationToken));
    Task<bool> WaitToReadAsync(CancellationToken cancellationToken = default(CancellationToken));
    Task Completion { get; }
}

public interface IWritableChannel<in T>
{
    bool TryWrite(T item);
    Task WriteAsync(T item, CancellationToken cancellationToken = default(CancellationToken));
    Task<bool> WaitToWriteAsync(CancellationToken cancellationToken = default(CancellationToken));
    bool TryComplete(Exception error = null);
}
```


### 4. System.Reactive / Rx

```C#
public interface IObservable<out T>
{
    IDisposable Subscribe(System.IObserver<T> observer);
}
public interface IObserver<in T>
{
    void OnCompleted();
    void OnError(Exception error);
    void OnNext(T value);
}
```

### 5. TPL Dataflow

``` C#
public interface ISourceBlock<out TOutput> : IDataflowBlock
{
    IDisposable LinkTo(ITargetBlock<TOutput> target, DataflowLinkOptions linkOptions);

    TOutput ConsumeMessage(DataflowMessageHeader messageHeader, ITargetBlock<TOutput> target, out Boolean messageConsumed);

    Boolean ReserveMessage(DataflowMessageHeader messageHeader, ITargetBlock<TOutput> target);

    void ReleaseReservation(DataflowMessageHeader messageHeader, ITargetBlock<TOutput> target);
}

public interface ITargetBlock<in TInput> : IDataflowBlock
{
	DataflowMessageStatus OfferMessage(DataflowMessageHeader messageHeader, TInput messageValue, ISourceBlock<TInput> source, Boolean consumeToAccept);
}
```


My thoughts about these are:

1. This is a pretty exact fit, but it is introducing a new API surface.

2. This is an existing API (good), but it's use of `StreamSequenceToken` is probably inappropriate for anything but the Streams API. (Perhaps there is a way to unify this concept with the Streams API...?)

3. This is the most flexible, but I think the `WaitTo*Async` calls are not needed. This API is not established yet and might change, therefore it might be too soon to adopt it. On the other hand, Orleans could potentially influence it's design and it might be a big win for everyone.

4. Rx is very well established and therefore we could tie in to a very broad set of existing API extensions that would be a huge win. But, Rx doesn't expose Tasks and therefore might not be a good fit. I haven't thought this one out fully. 

5. I think we can rule out exposing the actual DataFlow surface area since it is extremely specific.


Thinking about Rx (option 4) a bit more,it could work if this API doesn't need to provide guaranteed delivery. The producer could fire and forget an event and the subscriber need not provide response confirmation via the Task. But a big drawback of this is that `OnNext` would be a void returning method which gives Exceptions no logical place to go and perhaps worse if the subscriber needs to `await` something. Anything involving `async void` is probably a non-starter :(

ReubenBond commented:
Thank you, @AlgorithmsAreCool!

Some concerns with applying the **current** `IAsyncObservable`/`IAsyncObserver` interfaces to this PR:
* Operations like filtering, seeking, etc, should be provided before obtaining the observable rather than in the `SubscribeAsync` call. We cannot predict or cleanly specify the ways which users may want to query the underlying stream, and those query methods need to be discoverable. For example, maybe the stream doesn't support seeking (maybe it's hot), maybe it supports filtering by prefix, suffix, `string.Contains(x)`, or some complex query. Pushing these requirements to the point where the user obtains the stream allows for much more flexibility. Eventually, maybe we'll have `IAsyncQbservable<T>` and can support shifting expression trees to the most logical point in the cluster.
* `Subscribe` returns an `StreamSubscriptionHandle<T>` which is generic because it supports re-subscription via `ResumeAsync`. I believe resubscription should just be a `SubscribeAsync` call on the original observable. There is no `PauseAsync` method anyhow, only `UnsubscribeAsync`. For example: what's the difference between calling `StreamSubscriptionHandle<T>.ResumeAsync(...)` and calling `IAsyncObservable<T>.SubscribeAsync(...)`?
* `Subscribe` takes a `StreamSequenceToken`, which is cardinal (can be compared with other `StreamSequenceToken`s to determine which one is newer). This doesn't fit for ad-hoc streams, since they may have been created only for that specific call and each one is necessarily unique.

These details may be beneficial for Virtual Streams, but they are an awkward fit for these transient streams.

Maybe the Virtual Streams API can be simplified to make it more broadly applicable without throwing the baby out with the bath water. What are your opinions?

BTW: One caveat of my implementation is that the remote grain is only called when the stream is actually subscribed to. That means that any parameter validation by the remote grain is lazy. On the other hand, parameter validation errors for calling `Task<T>` grain methods are only seen once the returned task is `await`ed. So it's consistent.

gabikliot commented:
I think all your concerns re `IAsyncObservable/IAsyncObserver` can/should be solvebale.
The extra `Subsribe` method can be moved to a sub class: we can create a hierarchy of `IAsyncObservables` (plain, rewinbale, resumable, what ever).

The difference between resume and subscribe is described here: http://dotnet.github.io/orleans/Orleans-Streams/Streams-Programming-APIs. I think I can safely say that neither of us (me, @jason-bragg and @sergeybykov) like that extra API (Resume). We think its ugly, but it was the only way we could think of to solve the problems we had to solve.

The main point is (and I hope again that I am expressing everyone's opinion, please correct me if I am wrong): we all think the current streaming API is version 0.7 at best. The API should be improved: the internal implementation is rather robust I would say, but the API is first cut, and we need to improve it, to support all these other scenarios.

At the base it is very simple: it is identical to System.Rx just async. So fundamentally I don't think we made a wrong decision . But coming up with one unified API for all the cases was very hard, and this is work in progress.

I am of a strong opinion that instead of coming up with yet another API (your grain observers) we should improve the current one. And this where the help of the community, your help, @yevhen's help,  would be highly appropriated.

Another option is to give up completely on RX, both in virtaul and ad hoc streams and go with something else. I am not an expert on .NET Core but it seems like its new libraries are still immature as of now to be a strong candidate instead of Rx. On the other hand we were promised Rx 3.0, including async support, many years ago, and we are still waiting.

So my suggestion is: lets look at how to evolve the current `IAsyncObservable/IAsyncObserver` to be more fluent, user friendly and support all the different cases. This way we can also maybe impact the future System.Rx.

ReubenBond commented:
Sounds good. I agree with all of your points, Gabi. Let's keep this PR on
hold and open an issue to discuss modifications to the existing API.
Hopefully we can bring some of the Rx team/community in, too, in hopes of
aligning with them.
On 10 Jul 2016 2:16 PM, "Gabriel Kliot" <notifications@github.com> wrote:

> I think all your concerns re IAsyncObservable/IAsyncObserver can/should
> be solvebale.
> The extra Subsribe method can be moved to a sub class: we can create a
> hierarchy of IAsyncObservables (plain, rewinbale, resumable, what ever).
>
> The difference between resume and subscribe is described here:
> http://dotnet.github.io/orleans/Orleans-Streams/Streams-Programming-APIs.
> I think I can safely say that neither of us (me, @jason-bragg
> <https://github.com/jason-bragg> and @sergeybykov
> <https://github.com/sergeybykov>) like that extra API (Resume). We think
> its ugly, but it was the only way we could think of to solve the problems
> we had to solve.
>
> The main point is (and I hope again that I am expressing everyone's
> opinion, please correct me if I am wrong): we all think the current
> streaming API is version 0.7 at best. The API should be improved: the
> internal implementation is rather robust I would say, but the API is first
> cut, and we need to improve it, to support all these other scenarios.
>
> At the base it is very simple: it is identical to System.Rx just async. So
> fundamentally I don't think we made a wrong decision . But coming up with
> one unified API for all the cases was very hard, and this is work in
> progress.
>
> I am of a strong opinion that instead of coming up with yet another API
> (your grain observers) we should improve the current one. And this where
> the help of the community, your help, @yevhen <https://github.com/yevhen>'s
> help, would be highly appropriated.
>
> Another option is to give up completely on RX, both in virtaul and ad hoc
> streams and go with something else. I am not an expert on .NET Core but it
> seems like its new libraries are still immature as of now to be a strong
> candidate instead of Rx. On the other hand we were promised Rx 3.0,
> including async support, many years ago, and we are still waiting.
>
> So my suggestion is: lets look at how to evolve the current
> IAsyncObservable/IAsyncObserver to be more fluent, user friendly and
> support all the different cases. This way we can also maybe impact the
> future System.Rx.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/orleans/pull/1916#issuecomment-231569801>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAMcP7uwoiHOvp02ZBLIrI3jXPOA9fqqks5qUHIvgaJpZM4JHwQA>
> .
>

jdom commented:
> Thinking about Rx (option 4) a bit more,it could work if this API doesn't need to provide guaranteed delivery. The producer could fire and forget an event and the subscriber need not provide response confirmation via the Task. But a big drawback of this is that OnNext would be a void returning method which gives Exceptions no logical place to go and perhaps worse if the subscriber needs to await something. Anything involving async void is probably a non-starter :(

Well, the OnNext returning void is not really an issue, as I mentioned there are ways to turn this into a new grain request. One thing I'd like to clarify, as we might be talking about different things, is that I see returning an IObservable as a way to return a multi-valued response from a grain (that could be very short or very long lived). If that is the case, then I would make the producer be fire and forget by design (when you return a Task you get no feedback whether the original caller was able to handle it successfully). Observables are a great way to do this. Fire & forget doesn't mean messages in the middle could to be lost without either side's knowledge. For example, we could make sure all the events arrive in order to the observer. If a message is lost, we can send an OnError to the observer and unsubscribe it from the producer. The consumer can then decide to resubscribe.
In contrast to virtual streams, these observables are ephemeral (after all, it's just a multi-valued response, like returning a list of values at different times), so pausing/resuming, back-pressure, exceptions going back to the producer, should not be appropriate in a multi-valued response. It is the consumer who decides how to process that multi-valued response (honoring single-threadedness by default, and with an easy way to process each value as a new non-reentrant request). 
If the consumer cannot continue processing it can unsubscribe (that's the feedback loop in Rx... any other thing should be explicitly designed outside the API, such as back-pressure, in the same way as you would have to design how to stop a Task response prematurely, ie. with the usage of cancellation tokens in the original method call, it's not like you could tell the Task itself to stop producing a value and that automatically give feedback to the producer of that task value.

Are you guys thinking the same thing when you are designing this feature or did you have a different scenario in mind? Because with this view, I see no reason why the producer would ever have to await when calling OnNext, as long as we guarantee the multi-valued response is sent in order or ends in error (a similar guarantee we do for a Task response, we do not tamper with the payload and we do not guarantee the response will reach the original caller, and in that case the Task ends up faulted)

ReubenBond commented:
Opened #1931 for further discussion.

@jdom I mostly agree that grains usually don't care if the caller received the values pushed to it (as long as we let each end know of remote failures eventually).

Back-pressure may be desireable, but you're right that a similar concept to `CancellationToken` can be used during subscription in order for the caller to send its intent to the caller.

One issue with `IObservable<T>` is that `Subscribe` returns an `IDisposable`, which is a synchronous API. Another issue is that the API doesn't have a totally natural fit to Grains, which always run in the context of a task. I agree that we can create adapters, it just feels awkward.

sergeybykov commented:
>The main point is (and I hope again that I am expressing everyone's opinion, please correct me if I am wrong): we all think the current streaming API is version 0.7 at best. The API should be improved. The internal implementation is rather robust I would say, but the API is ugly, and we need to improve it, and to support all these other scenarios.

I second @gabikliot that the existing streaming API should not be viewed as a blocker for exploring better options. We entertained an idea of Streams v2 API, which would be defined from scratch with potentially drastically different programming model and leaving the existing Streaming API mostly as is as legacy.

The ideal solution would be to have a single coherent API that would cover all of these use cases. I don't know if that's possible though. Ideas are more than welcome.

ReubenBond commented:
Are there some Rx/Cortana team members who we can ask about the next Rx release or at least the API for the next release? Maybe @bartdesmet?

jdom commented:
> Back-pressure may be desireable, but you're right that a similar concept to CancellationToken can be used during subscription in order for the caller to send its intent to the caller.

Back-pressure is a complex topic. Solving it with OnNextAsync is not necessarily the right call in every situation anyway. There's some other alternatives in [Reactive Streams](https://github.com/reactive-streams/reactive-streams-dotnet) although not familiar with it, but might be worth considering.
Also, there is a very interesting topic on back-pressure in Rx: https://github.com/Reactive-Extensions/Rx.NET/issues/19

> One issue with IObservable<T> is that Subscribe returns an IDisposable, which is a synchronous API. 

This one is an interesting one. TBH I never liked that they coupled themselves to IDisposabe just because it helps when using the `using` keyword. Maybe there are already some alternatives for async? Nevertheless we could make the same guarantees we've been talking about here, and have the unsubscribe (dispose) call be best effort to be sent immediately (but the observer does not wait for the remote message to complete), and it will eventually reach the producer, either directly or because there is a network failure (in the same way that a disconnect between the 2 or a message being lost results in the observer receiving OnError and the producer an unsubscribe to the disposable).

> Another issue is that the API doesn't have a totally natural fit to Grains, which always run in the context of a task. I agree that we can create adapters, it just feels awkward.

This is probably not as awkward as you think. If you just want a similar API as the OnNextAsync for processing, you can just do `observable.ToAsyncEnumerable()` and have exactly the same code as you created in your examples. But as a benefit is that you could ALSO have each value be processed as a new grain request without reentrancy, which is not something I can see in the IGrainObservable example, unless I'm missing something. In your examples I do not see how you are dealing with reentrancy or asynchronous processing, I'm only seeing that you buffer all the values and when you think you are done receiving then you process all of them in one go ([in the case of the test client being the observer](https://github.com/dotnet/orleans/pull/1916/files#diff-043a39358bd045972092fbbd91e25499R25)) or you [mutate the list of chat messages](https://github.com/dotnet/orleans/pull/1916/files#diff-202f8db69c590157d7f0089ee45a1c7fR69) potentially violating reentrancy rules for that grain.
The point here is that with Rx you could choose what to do, with *familiar* concepts (at least familiar to people using Rx).

veikkoeeva commented:
This is a reference to authority, but this comes up time to time on Rx.NET too. The latest account: https://gitter.im/Reactive-Extensions/Rx.NET?at=577cb4cfddcdcee623fc3d53. The core issue is unbounded buffering, for what I gather, and then the difficulties implementing it so it won't break other operators. To cut to the core point of the discussion linked by @jdom: https://github.com/Reactive-Extensions/Rx.NET/issues/19#issuecomment-93260898.

ReubenBond commented:
There are no buffers in this PR, the buffer @jdom is referring to is just a helper class in the test code for convenience (i.e, what were all the values which were received by this observable). At least, I think that's what @jdom is referring to.

ReubenBond commented:
Latest revision eschews `IGrainObserver`, etc, in favor of the existing `IAsyncObservable`/ `IAsyncObserver`/`StreamSubscriptionHandle` types in order to align with Virtual Streams. It adds support for resuming streams after deactivation.

The `IAsyncObserver<T>`, `IAsyncObserver<T>` and `StreamSubscriptionHandle<T>` implementations are serializable, so each of them can be stored and reused later. Those entities cannot be transferred between grains/clients yet. So you cannot yet pass observers/observables/subscriptions between grains as arguments. I don't plan to implement that in this PR.

If a grain subscribes to another grain, it will be reactivated when it receives a new message. Similarly to Virtual Streams, grains need to call `ResumeAsync(...)` to resume the subscription. If the producer was deactivated, it will be reactivated at that point.

Calling `ResumeAsync(...)` on the `StreamSubscriptionHandle<T>` locally results in a subscribe on the remote grain's `IObservableGrainExtension`, passing in the existing `streamId`. If the remote grain does not hold a reference to the `StreamSubscriptionHandle` matching that `streamId`, then the grain's method will be called and a new subscription will be started with the existing `streamId`. In order to implement this, the `StreamSubscriptionHandle<T>` implementation holds a reference to the `IAsyncObservable<T>` impl, which holds a reference to the `InvokeMethodRequest` for the original grain method call.

@jdom wrote:
> But as a benefit is that you could ALSO have each value be processed as a new grain request without reentrancy, which is not something I can see in the IGrainObservable example, unless I'm missing something. In your examples I do not see how you are dealing with reentrancy or asynchronous processing, I'm only seeing that you buffer all the values and when you think you are done receiving then you process all of them in one go (in the case of the test client being the observer) or you mutate the list of chat messages potentially violating reentrancy rules for that grain.

This system works via grain extensions, which are processed in the context of a grain call. Orleans doesn't control client, but I *think* calls to client objects are invoked in a single-threaded context. Does anybody know for sure?

The feedback is very useful, keep it coming :smile:

jason-bragg commented on src/Orleans/Streams/AdHoc/IGrainObserver.cs:
```diff
@@ -0,0 +1,458 @@
+    /// <summary>
+    /// Represents a resource which can be disposed of asynchronously.
+    /// </summary>
+    public interface IAsyncDisposable
```
Suggest IObserverGrain.  Grain observer sounds like we're observing a grain, whereas this is a grain that observes something, hence IObserverGrain.

Similar suggestion for IGrainObservable => IObservableGrain

jason-bragg commented:
I started playing with this last night and a bit more tonight, mostly prototyping to get my head around the problem, but one thing I think we need to be very clear about here is our goals.  We got ourselves in trouble with the streaming apis because we modeled off of RX when what we were really building was a distributed real-time recoverable event processing framework (mind you, we didn't really know that at the time), which is -very- different from an asynchronous version of RX.

If we're looking to make it easier for developers to use RX with Orleans, even across grain boundaries, that sounds great.  It’s a clear goal and we know we should behave like RX and deviate only when dealing with the asynchronous nature of a distributed system.

However, if we’re looking at some other form of distributed observable infrastructure, I’d be very worried as we already have two; Observer grains and streams.

ReubenBond commented:
@jason-bragg I want the latter - a simple way to create and consume observable streams of data. By "Observer grains", I assume you mean `IGrainObserver`? If that is what you mean, or if you mean holding  calling references to other grains, then that's not the same. There are many times when I want to set up a user-defined standing query, such as:
* Show me the Pokemon near me (some lat/long)
* Show me *nearby* tweets tagged with #PokemonGO from last Wednesday
* Show me toasters in Melbourne which cost less than $20 and are gold-plated

These don't fit naturally with the existing virtual streams because they are transient queries - how would a producer know to start producing to that stream, for example, and how does a consumer find the stream? Would you encode the query into the stream id? 

Virtual streams are a powerful abstraction, but sometimes I want logical streams instead - particularly for 1:1 scenarios.

jason-bragg commented:
> " observable streams of data"

This is a very overloaded phrase...
In very simple terms it seems you want to:
-	Be able to ask another grain for data related to some sort of query that may not be immediately available.
-	Instead of receiving the data in the response, obtaining the data as it is available.
-	Have some means of knowing when no more data is available.

The reason I broke it down to the above statements, is that when dealing with observables, streams, and Rx, there are a number of expectations that can cloud and confuse the conversation.  When it comes to these types of processing models, there are a number of competing technologies that all have different advantages.  Reactive Extensions, TPL data flows, come to mind, but others were listed earlier in this thread.  As an actor model, I would discourage adding to these models, but instead allow integration with them.  This means providing grain call patterns and infrastructure that support these behaviors, then building model specific façade exposing these capabilities in a way that adheres (as much as is practicable) to specific models.

As an example, imaging we supported an azure storage like continuation token to our calls.  Something like.

    public class PartialResult<T>
    {
        T Results,
        Object continue;
    }

    Task< PartialResult<List<int>> GetAges(string filter, object continue = null);

A caller could make the request, then if the continuation was set, call for more data.  This is a pull model that could be wrapped by an RX façade to allow callers to do something like:

    {
        IDisposable disposer = Observable.Create(contineue => usersGrain.GetAges( “Name=Jim”, continue)) 
                                                                        .Subscribe(AverageAges);
    }

The advantage of this is that the Rx façade would work as Rx users expected, with no recoverability, no rewinding, and other such limitations of the Rx model, but the underlying api allows full access to advanced users to handle exceptions, retry operations, attempts to recover from transient errors, and the like, when necessary.

Additionally, the Rx façade could be part of an OrleansReactiveExtensions assembly that allows Orleans users to opt into these capabilities (and the related dependencies)

A similar façade could be built for TPL dataflows, or any other programming model that becomes popular.

If the pull model is not appropriate for all situations, a push model can be introduced and the façades can be updated to wrap it as well, allowing users to choose either push or pull models while maintaining the preferred higher level abstractions.

I am not actually suggesting the above APIs, I'm suggesting only the approach that we identify the specific underlying behavior we want to support; push models, pull models, call patterns, and the like, and build those, without regards to higher level programming abstractions like Rx, **then** build facades for specific programming models.  I'm suggesting an approach, not api or implementation specifics.

jason-bragg commented:
As this thread also seems to be the continuation of “Simplify IAsyncObservable/IAsyncObserver API #1931”, which I am of the opinion is still a useful discussion on its own, I offer the following on IAsyncObservable/IasyncObserver and future streaming apis.

I am of the opinion that IAsyncObservable/IasyncObserver should mirror as closely what the RX team prototyped for these which are simply async versions of the observable/observer apis.  This means no filtering or recovery information in these interfaces, and subscribe calls return asynchronously disposable objects.

I also don’t think the IAsyncObservable/IasyncObserver stream processing api pattern should be limited to stream providers.  We should be able to get streams from providers or grains.

This means we’ve a number of issues we’d need to solve:
-	How does recovery work?  How do you ask for a data from a particular point in a stream and how do you know what point to ask for?
-	How does filtering work?

I favor the following patterns:
-	When one requests a stream one gets an asynchronous version of a connectable observable.  Any filtering must be passed in the request for the observable.  Any stream position information must also be passed in this request.
-	If a stream is recoverable, it must pass any recoverability information along with the event.

For instance, a non-recoverable stream from sms stream provider would look like this.

    IAsyncConnectableObservable<int> subscription = sms.GetStream<int>(streamId);
    IAsyncDisposable handlerDisposer = await subscription.SubscribeAsync(ProcessSomeStuff, HandleError, Done);
    IasyncDisposable SubscriptionDisposer await subscription.Conect();

Same pattern would exist for a grain.

Recoverable streams from something like EventHub stream provider would look something like:

    IAsyncConnectableObservable<Recoverable<int>> subscription = eh.GetStream<int>(streamId, token);
    IAsyncDisposable handlerDisposer = await subscription.SubscribeAsync(ProcessSomeStuff, HandleError, Done);
    IAsyncDisposable SubscriptionDisposer await subscription.Conect();
 
Where Recoverable is a class that contains a value and a recovery token.  The nice thing about this is that we can generalize or specialize the token, and support different GetStream calls for different grains and stream providers.
For instance, the event hub stream provider could support the IRecoverableStreamProvider interface, which supports the generic recoverable calls as well as the IStreamProvider interface, which supports the non-recoverable calls (see the sms example), but could also support its own EventHub specific calls that users ok with the event hub dependency could use.
For instance:

    IAsyncConnectableObservable<EHRecoverable<int>> subscription = eh.GetEventHubStream<int>(streamId, token);

Where EHRecoverable has a value and an EventHub specific recovery token that has hub and position data that applications can use.

And/Or

    IAsyncConnectableObservable<EventData> subscription = eh.GetEventHubStream (streamId, token);

This call delivers EventHubs actual EventData with no interpretation.

ReubenBond commented:
@jason-bragg wrote:
> In very simple terms it seems you want to:
* Be able to ask another grain for data related to some sort of query that may not be immediately available.
* Instead of receiving the data in the response, obtaining the data as it is available.
* Have some means of knowing when no more data is available.

Yes, that's correct.

>As an actor model, I would discourage adding to these models, but instead allow integration with them. 

In the latest commits, I've aligned with the existing Virtual Streams API, so I'm not adding to those models.

> Additionally, the Rx façade could be part of an OrleansReactiveExtensions assembly that allows Orleans users to opt into these capabilities (and the related dependencies)
> A similar façade could be built for TPL dataflows, or any other programming model that becomes popular.

I agree - I think that we should create façades to support each alternative, but only as-desired.

> I also don’t think the IAsyncObservable/IasyncObserver stream processing api pattern should be limited to stream providers. We should be able to get streams from providers or grains.

Great :)

> This means we’ve a number of issues we’d need to solve:
* How does recovery work? How do you ask for a data from a particular point in a stream and how do you know what point to ask for?

This is at least partially solved: consumers (both clients & grains) can call `ResumeAsync`/`SubscribeAsync`, passing in the latest `StreamSequenceToken`. The semantics are specified in [this comment](https://github.com/dotnet/orleans/pull/1916#issuecomment-232331824). They are given the point to ask for in `OnNextAsync` calls. Alternatively they could make the resumption point a parameter to the method which returns the observable, eg:
```C#
IAsyncObservable<string> GetMessages(DateTime sinceTime);`
```
> * How does filtering work?

I'm not a fan of the `filterObject` approach used in Virtual Streams, but I understand its usefulness and the reasons why it came to be - it's just a little less necessary when consumers are connected more directly with producers. In the case of this PR I think we should propagate the `filterObject` value provided, but I expect that method parameters to control which values are produced (as above) will be more popular.

For the remainder of the comments, I believe the existing API is sufficient for the time being, but agree that we should seriously consider evolving the API to better express the behavior/guarantees of the different kinds of streams (eg, not all of them are seekable or reliable). I think I was wrong to intertwine these two issues, though.

Do you think the design of the streams API is a blocking issue for this PR?

jason-bragg commented:
> "Do you think the design of the streams API is a blocking issue for this PR?"

No, I think they are separate issues.  I am, however, less comfortable than @gabikliot with using the existing streaming APIs with this feature.  I am of the opinion that the existing stream APIs were based off a good idea, but now that we know more, they are not what I'd recommend.  We're not really in a good place here.  We can:
- Building this system using the existing async observ[er][able] interfaces and incur the hit, internally and to users, if/when we refactor/replace them.
- Build this system using new interfaces, and incur the cost and confusion of having two sets of interfaces.
- Hold off on this change until we refactor the stream interfaces, which means denying users this, rather nice, capability.

I slightly favor option 2, introducing new interfaces for IAsyncObserver and IAsnycObservable, maybe in different namespaces?  But this is only a slight preference.  I'm ok with using the existing streaming api's if that is the general consensus.

-------------

>This (recoverablity) is at least partially solved: consumers (both clients & grains) can call  ResumeAsync / SubscribeAsync , passing in the latest  StreamSequenceToken.

I apologize for being unclear, but I meant we'd have to solve these problems using the new apis.  Yes, the existing api's solve this, but they do this by polluting the asynchronous reactive interfaces.  What I was exploring was how we would solve this using the following interfaces:


    public interface IAsyncObservable<out T>
    {
        Task<IAsyncDisposable> SubscribeAsync(IAsyncObserver<T> observer, CancellationToken token = default(CancellationToken));
    }

    public interface IAsyncObserver<in T> : IAsyncDisposable
    {
        Task OnNextAsync(T value, CancellationToken token = default(CancellationToken));
        Task OnErrorAsync(Exception error, CancellationToken token = default(CancellationToken));
        Task OnCompletedAsync(CancellationToken token = default(CancellationToken));
    }

    public interface IAsyncDisposable
    {
        Task DisposeAsync(CancellationToken token = default(CancellationToken));
    }

and possibly and IAsyncConnectableObservable..

jason-bragg commented:
@ReubenBond, 
>"I think I was wrong to intertwine these two issues, though."

I'd not argue you were wrong, as much as by doing so the discussion is wider and more difficult to solve.  This is in part why I attempted to identify the specific behaviors you needed.  I was hoping to narrow the problem in an attempt to reach a consensus.  I am now of the opinion that since then I've hindered my own effort by providing comments on “Simplify IAsyncObservable/IAsyncObserver API #1931”.

I'll attempt to narrow this again by reopening #1931, for stream api discussions, and discussing only the immediate requirements here.  I hope this is not taken badly, as I'm not trying to avoid the discussion as much as focus on the specific needs instigating this PR.

To restate the requirements so far gathered:
-	Be able to ask another grain for data related to some sort of query that may not be immediately available.
-	Instead of receiving the data in the response, obtaining the data as it is available.
-	Have some means of knowing when no more data is available.

Historically, patterns I’m aware of for this type of behavior fit into one of two rough categories:
-	Push – Client makes request for data, and server sends it.
-	Pull – Client makes repeated requests for data.

I raise these models because they’ve been around for a while and are mostly understood, especially as related to things like error handling and responsibilities.  As such, my preference would be to pick one of these patterns then build facades to fit higher level abstractions like Rx or TPL DataFlows.   I suggest this because I’m of the opinion we are more likely to get these well understood patterns correct than we would be shooting for something like an asynchronous version of reactive extensions.  This is of course not as interesting, but (as you’ve likely noticed) I’m more risk averse than others, so I favor simple, well understood patterns.

This changes the question to:
-	Do we want to use a push pattern or a pull pattern?
-	How do we setup the selected pattern between grains?

I am not adamantly opposed to the approach used in this PR, I simply don’t have high confidence in our ability to reason about responsibilities and/or error handling, hence (IMO) we’re adding a model we’re less likely to get right.

ReubenBond commented:
> Do we want to use a push pattern or a pull pattern?

Push. We can implement pull at another time if people want it.
> How do we setup the selected pattern between grains?

I hope I understand the question. I like the way this PR does it: expose streams via grain methods. It's statically typed, discoverable, and it's how users would naturally do it in a non-distributed context. It also allows us to expose different patterns (`IAsyncEnumerable<T>`, `IProducerBlock<T>`) later as desired.

> I am not adamantly opposed to the approach used in this PR, I simply don’t have high confidence in our ability to reason about responsibilities and/or error handling, hence (IMO) we’re adding a model we’re less likely to get right.

I'm not adding a model, just using the existing one.

For error handling, can you see any gaps here?
- `IAsyncObserver<T>`, `IAsyncObservable<T>`, `StreamSubscriptionHandle<T>`, & `StreamSequenceToken` instances from the framework can be serialized and deserialized and used by reactivated grains.
- `OnNextAsync`/`OnErrorAsync`/`OnCompletedAsync`:
    - Reactivates the consumer if it's a grain.
    - Throws if the consumer could not be reactivated (for example, if it's a client which has since failed).
        - The producer can decide whether or not to remove the observer.
- `ResumeAsync`:
    - If the producer has been deactivated, it will be reactivated and the original (`IAsyncObservable<T>`) method will be called again.
        - It is equivalent to a call to `SubscribeAsync` in this case.
    - If the producer is still active, it will NOT be notified at all, but the registration for that `IAsyncObservable<T>` will be updated to point to the new consumer.
        - This allows for non-grain consumers to recover subscriptions, whereby the consumer's address will have changed.
        - This also allows for subscriptions to be transferred between consumer grains.
        - (TODO) `ResumeAsync` is called on the `StreamSubscriptionHandle<T>` returned from `SubscribeAsync`.
    - If the caller was activated by an `OnNextAsync`/`OnErrorAsync`/`OnCompletedAsync` call, calling `ResumeAsync` in the grain's `OnActivateAsync` method allows the current call to succeed without error.
- `UnsubscribeAsync`:
    - (TODO) If the producer has been deactivated, it will be reactivated and `UnsubscribeAsync` will be called on the `StreamSubscriptionHandle<T>` returned from `SubscribeAsync`.
        - NOTE: currently this is a no-op, meaning that if the producer persisted the `IAsyncObserver<T>` instance, calls to unsubscribe will not be propagated and the producer will only be informed once they attempt to message that observer (calls to it will throw).
    - If the producer is still active, `UnsubscribeAsync` is called on the `StreamSubscriptionHandle<T>` returned from `SubscribeAsync`.

There are some issues/questions with the current approach:
1. Should `ResumeAsync` be equivalent to `SubscribeAsync` on a reactivated producer?
2. Should `ResumeAsync` & `UnsubscribeAsync` calls be propagated to a reactivated producer?
3. If so, does the framework handle observer management (including persistence) or do we allow grains to register persisted `IAsyncObservable<T>`/`StreamSubscriptionHandle<T>` instances manually?

My preferred answers are: yes, no, and N/A respectively. I would understand differing opinions.
My reasoning is that it's a scalable model, since:
    - Consumers are responsible for storing `StreamSubscriptionHandle<T>` instances themselves and can use those to resume/stop subscriptions.
    - Producers are not responsible for storing anything, but can store `IAsyncObserver<T>` instances if they wish. If the producer tries to call methods on an unsubscribed observer, an exception will be thrown and the producer can remove the observer. TODO: Make the exception well-typed.

ReubenBond commented:
Is evolution of Virtual Streams needed before this PR/#940? Should such an initiative come from the core team or should an outside contributor be preparing a proposal?

jason-bragg commented:
This is a good question...

I've spoken with the team here about our roadmap and the future of streaming.  I expect we will improve our planning in the following weeks.  We have a number of internal (and external) users of the current implementations, meaning that I need to keep focused on completing and hardening these systems, despite the fact that my experiences with them have convinced me that an evolution of the patterns is needed.  In short, I'd very much like to run forward as fast as I can, but my priorities are the current systems used in shipp[ed|ing] products.  This leads me to suspect that virtual streams II will need to be spearheaded by the community or delayed until the existing systems are completed + hardened.

Regarding this feature, If we build this feature using the existing streaming interfaces, we burden it with all of the problems with the current virtual streams system and further entrench those patterns.  If we do something different, we introduce yet another model.  If we wait until virtual streams II, we forgo this feature for some undetermined time.

ReubenBond commented:
Given your experience with Virtual Streams v1, what are the problems we should address in v2? My appreciation for the current API increased after playing with it and implementing recovery

ReubenBond commented:
Will reopen when I make further progress
